src/virtio_queue/split_queue.rs

```diff
@@ -211,13 +319,33 @@ impl<H: Hal, const SIZE: usize> VirtQueue<H, SIZE> {
     /// Returns whether the driver should notify the device after adding a new buffer to the
     /// virtqueue.
     ///
+    /// Ref: virtio_ring.c virtqueue_kick_prepare_split
     /// This will be false if the device has supressed notifications.
-    pub fn should_notify(&self) -> bool {
+    pub fn should_notify(&self, support_event: bool) -> bool {
         // Read barrier, so we read a fresh value from the device.
         fence(Ordering::SeqCst);
 
         // Safe because self.used points to a valid, aligned, initialised, dereferenceable, readable
         // instance of UsedRing.
+        if support_event {
+            unsafe { (*self.avail.as_ptr()).idx == (*self.used.as_ptr()).avail_event }
```

Comment:
This should use `self.avail_idx` rather than reading from the available ring directly, because we don't trust the device not to modify the available ring (even though it shouldn't).

Also, from my reading of the spec, it sounds like we should be checking for `avail_idx == avail_event.wrapping_add(1)`, as `should_notify` is called after `avail_idx` is updated. See 2.7.10.2: "equivalently, until idx in the available ring will reach the value avail_event + 1".

Reply:
I'll send a separate PR to add event_idx support with this fixed.

Reply:
Let me read the spec again to figure it out